// convex/dist/esm/cli/lib/ai/config.js
"use strict";
import * as Sentry from "@sentry/node";
import { promises as fs } from "fs";
import path from "path";
import { z } from "zod";
import { aiFilesStatePathForConvexDir } from "./paths.js";
const aiFilesStateSchema = z.object({
  guidelinesHash: z.string().nullable(),
  agentsMdSectionHash: z.string().nullable(),
  claudeMdHash: z.string().nullable(),
  // Commit SHA from get-convex/agent-skills that was current when skills were
  // last installed. Used to detect when newer skills are available.
  agentSkillsSha: z.string().nullable(),
  // Names of skills installed by `npx skills add`, used by `remove` to
  // only remove Convex-managed skills.
  installedSkillNames: z.array(z.string()).default([])
});
const aiFilesProjectConfigSchema = z.object({
  aiFiles: z.object({
    disableStalenessMessage: z.boolean().default(false)
  }).default({ disableStalenessMessage: false })
}).passthrough();
export const aiFilesSchema = aiFilesStateSchema;
const EMPTY_AI_STATE = {
  guidelinesHash: null,
  agentsMdSectionHash: null,
  claudeMdHash: null,
  agentSkillsSha: null,
  installedSkillNames: []
};
async function readAiDisabledFromProjectConfig(projectDir) {
  let raw;
  try {
    raw = await fs.readFile(path.join(projectDir, "convex.json"), "utf8");
  } catch {
    return false;
  }
  try {
    const parsed = aiFilesProjectConfigSchema.parse(JSON.parse(raw));
    return parsed.aiFiles.disableStalenessMessage;
  } catch (err) {
    Sentry.captureException(err);
    return false;
  }
}
export async function writeAiDisabledToProjectConfig(disableStalenessMessage, projectDir) {
  const filePath = path.join(projectDir, "convex.json");
  let existing = {};
  try {
    existing = JSON.parse(await fs.readFile(filePath, "utf8"));
  } catch {
  }
  const base = existing !== null && typeof existing === "object" && !Array.isArray(existing) ? existing : {};
  const aiFilesValue = base.aiFiles !== null && typeof base.aiFiles === "object" && !Array.isArray(base.aiFiles) ? base.aiFiles : {};
  const { $schema, ...rest } = base;
  const next = {
    $schema: $schema ?? "node_modules/convex/schemas/convex.schema.json",
    ...rest,
    aiFiles: {
      ...aiFilesValue,
      disableStalenessMessage
    }
  };
  await fs.writeFile(filePath, JSON.stringify(next, null, 2) + "\n", "utf8");
}
export async function readAiConfig(projectDir, convexDir) {
  const disableStalenessMessage = await readAiDisabledFromProjectConfig(projectDir);
  let rawState;
  try {
    rawState = await fs.readFile(
      aiFilesStatePathForConvexDir(convexDir),
      "utf8"
    );
  } catch {
    return disableStalenessMessage ? { ...EMPTY_AI_STATE, disableStalenessMessage } : null;
  }
  try {
    const state = aiFilesStateSchema.parse(JSON.parse(rawState));
    return {
      ...state,
      disableStalenessMessage
    };
  } catch (err) {
    Sentry.captureException(err);
    return null;
  }
}
export async function writeAiConfig(config, projectDir, convexDir, options) {
  const state = aiFilesStateSchema.parse({
    guidelinesHash: config.guidelinesHash,
    agentsMdSectionHash: config.agentsMdSectionHash,
    claudeMdHash: config.claudeMdHash,
    agentSkillsSha: config.agentSkillsSha,
    installedSkillNames: config.installedSkillNames
  });
  await fs.writeFile(
    aiFilesStatePathForConvexDir(convexDir),
    JSON.stringify(state, null, 2) + "\n",
    "utf8"
  );
  const persistMode = options?.persistDisabledPreference ?? "ifTrue";
  if (persistMode === "always" || persistMode === "ifTrue" && config.disableStalenessMessage) {
    await writeAiDisabledToProjectConfig(
      config.disableStalenessMessage,
      projectDir
    );
  }
}
